packages/server/src/ratings.ts

```diff
@@ -0,0 +1,169 @@
+import {
+  GameResponse,
+  AbstractGame,
+  UserRanking,
+  UserRankings,
+  UserResponse,
+  GameResult,
+  GameResults,
+} from "@ogfcommunity/variants-shared";
+import { Glicko2, Player } from "glicko2";
+import { updateUserRanking, getUser, updateUserGameHistory } from "./users";
+
+const ranking = new Glicko2({
+  tau: 0.5,
+  rating: 1500,
+  rd: 350,
+  vol: 0.06,
+});
+
+const matches: [Player, Player, number][] = [];
+
+export async function updateRatings(
+  game: GameResponse,
+  game_obj: AbstractGame<object, object>,
+) {
+  const variant = game.variant;
+
+  let glicko_outcome;
+  let outcome_player_black;
+  let outcome_player_white;
+
+  if (game_obj.result.endsWith("R")) {
+    if (game_obj.result[0] === "B") {
+      glicko_outcome = 0;
+      outcome_player_black = "Loss";
+      outcome_player_white = "Win";
+    } else if (game_obj.result[0] === "W") {
+      glicko_outcome = 1;
+      outcome_player_black = "Win";
+      outcome_player_white = "Loss";
+    }
+  } else {
+    if (game_obj.result[0] === "B") {
+      glicko_outcome = 1;
+      outcome_player_black = "Win";
+      outcome_player_white = "Loss";
+    } else if (game_obj.result[0] === "W") {
+      glicko_outcome = 0;
+      outcome_player_black = "Loss";
+      outcome_player_white = "Win";
+    } else {
+      glicko_outcome = 0.5;
+      outcome_player_black = "Tie";
+      outcome_player_white = "Tie";
+    }
+  }
+
+  const player_black_id = game.players[0].id;
+  const player_white_id = game.players[1].id;
+
+  const db_player_black = await getUser(player_black_id);
+  const db_player_white = await getUser(player_white_id);
+
+  if (db_player_black.ranking == undefined) {
+    db_player_black.ranking = {};
+  }
+  if (db_player_white.ranking == undefined) {
+    db_player_white.ranking = {};
+  }
+
+  const new_game_history_player_black = updatedGameHistory(
+    db_player_black,
+    db_player_white.ranking[variant],
+    outcome_player_black,
+    variant,
+  );
+  const new_game_history_player_white = updatedGameHistory(
+    db_player_white,
+    db_player_black.ranking[variant],
+    outcome_player_white,
+    variant,
+  );
+
+  await Promise.all([
```

Comment:
I see. I thought you meant for game history we store the matches that the user has played simply for the purposes of keeping track what games a user has played. I get what you mean now.

Reply:
Now that I think about it, do we really need a history? The third DB hit is coming from updating the history. Instead, if we pull the ranking and then update the ranking back to the DB that would take 2 hits. And since we are updating ranking after each game we wouldn't need a history.

Even if we really need to include the history we can use 1 DB hit to update both the ranking and the history at the same time.  Eg:
```
const update_result = await usersCollection().updateOne(
    { _id: new ObjectId(user_id) },
    { $set: { ranking: new_ranking, gameHistory: new_game_history } },
  );
```

Reply:
Yeah, this was in response to your question about where to store a "matches" array:

> Currently, the user ratings are updated after every match, which is not ideal according to the glicko2 library. Instead, after a match, I want to add the result to the matches array and then store the array somewhere in the database. Then when the matches array has 10 elements, we can update the ratings and clear the array. So my question is where in the database should I store the matches array?

To be clear, this suggestion is assuming we can process one player at a time.  As described in the glicko2 docs ["How to deal with a big database of players"](https://github.com/mmai/glicko2js?tab=readme-ov-file#how-to-deal-with-a-big-database-of-players):

> ~At the end of the week~ *When we reach 10 matches for a player*, ~for each player (even those that did not play)~:
>
> - load from the database:
>     - the player *'s current* ranking data ~for the current week~
>     - the matches the player has played ~during the week~ *(from game history)*
>     - for each match, the player's opponent data ~for the current week~ *(also stored in game history)*
> - add the player and his opponents to a new glicko instance
> - add the matches to the glicko instance
> - update glicko rankings
> - save the player updated ranking data to database as the ~next week~ *new* player data

Reply:
just saw your new comment - no we don't need history if we update ratings one by one! Feel free to remove it and we can figure out how to improve things with a game history later.

Reply:
Got it!